### ISSUE TITLE
Caused by: java.lang.ClassCastException: class com.android.build.gradle.internal.plugins.AppPlugin$ApplicationAndroidComponentsExtensionImplCompat_Decorated cannot be cast to class kotlin.Unit

### ISSUE THREAD

**Issue by bacecek** [gradle/gradle#23148]
### Expected Behavior
Successful project configuration

### Current Behavior
Got build failure during project configuration. 
```
Caused by: java.lang.ClassCastException: class com.android.build.gradle.internal.plugins.AppPlugin$ApplicationAndroidComponentsExtensionImplCompat_Decorated cannot be cast to class kotlin.Unit (com.android.build.gradle.internal.plugins.AppPlugin$ApplicationAndroidComponentsExtensionImplCompat_Decorated is in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @48ee7c62; kotlin.Unit is in unnamed module of loader org.gradle.internal.classloader.VisitableURLClassLoader @548e7350)
	at Dev_bacecek_base_gradle.<init>(dev.bacecek.base.gradle.kts:31)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at Dev_bacecek_basePlugin.apply(Dev_bacecek_basePlugin.kt:13)
	at Dev_bacecek_basePlugin.apply(Dev_bacecek_basePlugin.kt:7)
	at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:43)
	at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:51)
	at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:187)
	at org.gradle.api.internal.plugins.DefaultPluginManager.access$100(DefaultPluginManager.java:52)
	at org.gradle.api.internal.plugins.DefaultPluginManager$AddPluginBuildOperation.run(DefaultPluginManager.java:282)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
	at org.gradle.api.internal.plugins.DefaultPluginManager.lambda$doApply$0(DefaultPluginManager.java:167)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.apply(DefaultUserCodeApplicationContext.java:44)
	at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:166)
	... 210 more
```

### Context
The stacktrace and code contains Android related code, but the main cause does not seems to be related to Android by itself. The issue is reproduced in combination by several factors: using kotlin `.apply` function under some condition and this block is the latest in the file. For example:
```
// buildSrc/src/main/kotlin/example.gradle.kts
...
if (true) {
    extensions.getByName<BaseExtension>("example").apply {
        // some code here
    }
}

//end of the file
```

So when I remove one of the factors (use `run` instead of `apply` or remove condition or move it above other code/add any code below) the issue is not respoducing.

### Steps to Reproduce 
Configure project using any task on module where the plugin is applied (for example in sample project `./gradlew :app:assembleDebug`) or sync project in Intellij IDEA/Android Studio.

### Your Environment
Gradle version 7.6. Also reproduced in 7.5.
Sample project to reproduce the issue: https://github.com/bacecek/sample-checking-assistant-role/tree/gradle/classcastexception_in_precompiled_script_plugin and precompiled script code here: https://github.com/bacecek/sample-checking-assistant-role/blob/gradle/classcastexception_in_precompiled_script_plugin/buildSrc/src/main/kotlin/dev.bacecek.base.gradle.kts.

**Comment by xaviarias** [comment]
This seems caused by a bug in Kotlin that has been fixed in 1.8, and so it doesn't reproduce in Gradle 8.0.

Closing as resolved.

**Comment by niffin** [comment]
hi @xaviarias -- do you mind linking to something related to the Kotlin bug that causes this?

**Comment by xaviarias** [comment]
Hi @niffin, unfortunately we don't have more details on the issue.

As far as we know, this was a Kotlin issue not related to Gradle, and it's been fixed in Kotlin 1.8.
